File: UINavigationExtensionDemo/Feature/Advanced/Controllers/ViewController06_ClickEventHitToBack.m
//
//  ViewController06_ClickEventHitToBack.m
//  UINavigationExtensionDemo
//
//  Created by Leo Lee on 2020/10/26.
//

#import <UINavigationExtension/UINavigationExtension.h>

#import "ViewController06_ClickEventHitToBack.h"
#import "UIColor+RandomColor.h"

@interface ViewController06_ClickEventHitToBack ()

@property (nonatomic, strong) UIView *tableHeaderView;
@property (nonatomic, strong) UISegmentedControl *segmentedControl;

@property (nonatomic, strong) NSLayoutConstraint *leftConstraint;
@property (nonatomic, strong) NSLayoutConstraint *rightConstraint;
@property (nonatomic, strong) NSLayoutConstraint *heightConstraint;

@end

@implementation ViewController06_ClickEventHitToBack

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = nil;
    self.tableView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    
    [self.tableHeaderView addSubview:self.segmentedControl];
    
    self.segmentedControl.translatesAutoresizingMaskIntoConstraints = NO;
    [self.segmentedControl.bottomAnchor constraintEqualToAnchor:self.tableHeaderView.bottomAnchor].active = YES;
    
    self.leftConstraint = [self.segmentedControl.leftAnchor constraintEqualToAnchor:self.tableHeaderView.leftAnchor];
    self.leftConstraint.active = YES;
    
    self.rightConstraint = [self.segmentedControl.rightAnchor constraintEqualToAnchor:self.tableHeaderView.rightAnchor];
    self.rightConstraint.active = YES;
    
    self.heightConstraint = [self.segmentedControl.heightAnchor constraintEqualToConstant:0];
    self.heightConstraint.active = YES;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];

    CGRect navigationBarFrame = self.navigationController.navigationBar.frame;
    CGFloat tableHeaderViewX = navigationBarFrame.origin.x;
    CGFloat tableHeaderViewY = navigationBarFrame.origin.y;
    CGFloat tableHeaderViewWidth = navigationBarFrame.size.width;
    CGFloat tableHeaderViewHeight = tableHeaderViewY + navigationBarFrame.size.height;

    self.tableHeaderView.frame = CGRectMake(tableHeaderViewX, tableHeaderViewY, tableHeaderViewWidth, tableHeaderViewHeight);
    self.tableView.tableHeaderView = self.tableHeaderView;
    
    UIEdgeInsets safeAreaInsets = self.navigationController.navigationBar.safeAreaInsets;
    self.heightConstraint.constant = navigationBarFrame.size.height;
    self.leftConstraint.constant = safeAreaInsets.left;
    self.rightConstraint.constant = -safeAreaInsets.right;
}

- (BOOL)ue_hidesNavigationBar {
    return YES;
}

- (BOOL)ue_enableFullScreenInteractivePopGesture {
    return YES;
}

#pragma mark - Getter

- (UIView *)tableHeaderView {
    if (!_tableHeaderView) {
        _tableHeaderView = [[UIView alloc] initWithFrame:CGRectZero];
        _tableHeaderView.backgroundColor = [UIColor groupTableViewBackgroundColor];
    }
    return _tableHeaderView;
}

- (UISegmentedControl *)segmentedControl {
    if (!_segmentedControl) {
        _segmentedControl = [[UISegmentedControl alloc] initWithItems:@[@"First", @"Second"]];
        _segmentedControl.selectedSegmentIndex = 0;
        [_segmentedControl setTitleTextAttributes:@{NSForegroundColorAttributeName: [UIColor customDarkGrayColor]} forState:UIControlStateNormal];
        [_segmentedControl setTitleTextAttributes:@{NSForegroundColorAttributeName: [UIColor customLightGrayColor]} forState:UIControlStateSelected];
        _segmentedControl.backgroundColor = [UIColor customLightGrayColor];
        [_segmentedControl addTarget:self action:@selector(changeSegmentedControl:) forControlEvents:UIControlEventValueChanged];
        if (@available(iOS 13.0, *)) {
            _segmentedControl.selectedSegmentTintColor = [UIColor customDarkGrayColor];
        } else {
            _segmentedControl.tintColor = [UIColor customDarkGrayColor];
        }
    }
    return _segmentedControl;
}

- (void)changeSegmentedControl:(UISegmentedControl *)segmentedControl {
    NSLog(@"%@", [segmentedControl titleForSegmentAtIndex:segmentedControl.selectedSegmentIndex]);
}

@end

File: Example/UNXNavigatorSwiftDemo/Advanced/ViewController06_ClickEventHitToBack.swift
//
//  ViewController06_ClickEventHitToBack.swift
//  UNXNavigatorSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit

class ViewController06_ClickEventHitToBack: BaseTableViewController {
    
    private lazy var tableHeaderView: UIView = {
        let view = UIView(frame: .zero)
        view.backgroundColor = .customGroupedBackground
        return view
    }()
    
    private lazy var segmentedControl: UISegmentedControl = {
        let segmentedControl = UISegmentedControl(items: ["First", "Second"])
        segmentedControl.selectedSegmentIndex = 0
        segmentedControl.backgroundColor = .customLightGray
        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
        segmentedControl.setTitleTextAttributes([NSAttributedString.Key.foregroundColor: UIColor.customDarkGray], for: .normal)
        segmentedControl.setTitleTextAttributes([NSAttributedString.Key.foregroundColor: UIColor.customLightGray], for: .selected)
        segmentedControl.addTarget(self, action: #selector(changeSegmentedControl(_:)), for: .valueChanged)
        if #available(iOS 13.0, *) {
            segmentedControl.selectedSegmentTintColor = .customDarkGray
        } else {
            segmentedControl.tintColor = .customDarkGray
        }
        return segmentedControl
    }()
    
    private var leftConstraint: NSLayoutConstraint?
    private var rightConstraint: NSLayoutConstraint?
    private var heightConstraint: NSLayoutConstraint?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        navigationItem.title = nil
        
        tableView.contentInsetAdjustmentBehavior = .never
        tableHeaderView.addSubview(segmentedControl)
        
        segmentedControl.bottomAnchor.constraint(equalTo: tableHeaderView.bottomAnchor, constant: -4).isActive = true
        
        leftConstraint = segmentedControl.leftAnchor.constraint(equalTo: tableHeaderView.leftAnchor)
        leftConstraint?.isActive = true
        
        rightConstraint = segmentedControl.rightAnchor.constraint(equalTo: tableHeaderView.rightAnchor)
        rightConstraint?.priority = .defaultHigh
        rightConstraint?.isActive = true
        
        heightConstraint = segmentedControl.heightAnchor.constraint(equalToConstant: 0.0)
        heightConstraint?.isActive = true
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        let navigationBarFrame = navigationController?.navigationBar.frame ?? .zero
        
        let tableHeaderViewX = navigationBarFrame.minX
        let tableHeaderViewY = navigationBarFrame.minY
        let tableHeaderViewWidth = navigationBarFrame.width
        let tableHeaderViewHeight = tableHeaderViewY + navigationBarFrame.height
        tableHeaderView.frame = CGRect(x: tableHeaderViewX, y: tableHeaderViewY, width: tableHeaderViewWidth, height: tableHeaderViewHeight)
        tableView.tableHeaderView = tableHeaderView
        
        let safeAreaInsets = self.navigationController?.navigationBar.safeAreaInsets ?? .zero
        heightConstraint?.constant = navigationBarFrame.height * 0.8
        leftConstraint?.constant = safeAreaInsets.left
        rightConstraint?.constant = -safeAreaInsets.right
    }
    
    override var unx_hidesNavigationBar: Bool {
        return true
    }
    
    override var unx_enableFullScreenInteractivePopGesture: Bool {
        return true
    }

}

@objc
extension ViewController06_ClickEventHitToBack {
    
    private func changeSegmentedControl(_ segmentedControl: UISegmentedControl) {
        print(segmentedControl.titleForSegment(at: segmentedControl.selectedSegmentIndex) ?? "")
    }
    
}
